### ISSUE TITLE
Add ability to move and sort between lists

### ISSUE THREAD

**Issue by denzo** [jgwhite/ember-sortable#38]
I couldn't figure out how to move and sort between lists.

**Comment by jgwhite** [comment]
ember-sortable doesn’t have this capability yet. @rlivsey has done some work in this area. We want the API to be simple and fit with the data-down-actions-up ethos. API proposals very very welcome :D

**Comment by rlivsey** [comment]
I've not done any work on this yet, but will need this functionality soon so I'll make a start if no-one else gets there first!

Thinking out loud here…

One API could be to simply call `onChange` on both lists with the new sets. 

One downside to that is that you have to infer that a move has happened by comparing the current lists to the new ones.

An alternative API would be to add a couple of new events like `item-removed` and `item-added` which could trigger with `item-added` including the item itself and the position to insert it at?

Another area of complexity is that currently on mouse move we simply loop through the list of items comparing their Y position to figure out the ordering. How do we know you've exited one list and entered a new one. We could loop through the union of all lists, but that seems excessive. I guess it could be easily culled by getting the bounding box of each list when the drag starts.

Next up, how do the lists know about each other? Is there a global registry (a sortable service?) that the lists register themselves with, much like how the items register themselves with the list at the moment?

**Comment by jgwhite** [comment]
As a starting point, I think groups should probably be nestable:

```hbs
{{#sortable-group onChange=(action "updateLists") as |parentGroup|}}

  {{#sortable-group group=parentGroup as |groupA|}}
    {{#each listA as |item|}}
      {{#sortable-item model=item group=groupA}}
        {{item.name}}
      {{/sortable-item}}
    {{/each}}
  {{/sortable-group}}

  {{#sortable-group group=parentGroup as |groupB|}}
    {{#each listB as |item|}}
      {{#sortable-item model=item group=groupB}}
        {{item.name}}
      {{/sortable-item}}
    {{/each}}
  {{/sortable-group}}

{{/sortable-group}}
```

```js
// Whichever class or route is in charge...

actions: {
  updateLists(/* what should this receive? */) {
    // what should it do?
  }
}
```

**Comment by rlivsey** [comment]
@jgwhite I like that. It would certainly work for my personal use-case.

Maybe dragging within one list would call the `onChange` of that list, but dragging between lists would call `onChange` of the parent group?

Will think about this some more…

**Comment by jgwhite** [comment]
I’m not sure what the parent’s `onChange` would send. Need to sketch out some use cases.

For child list `onChange` actions, I’m thinking in this case:

```hbs
{{#sortable-group as |parentGroup|}}

  {{#sortable-group group=parentGroup onChange=(action updateListA) as |groupA|}}
    {{#each listA as |item|}}
      {{#sortable-item model=item group=groupA}}
        {{item.name}}
      {{/sortable-item}}
    {{/each}}
  {{/sortable-group}}

  {{#sortable-group group=parentGroup onChange=(action updateListB) as |groupB|}}
    {{#each listB as |item|}}
      {{#sortable-item model=item group=groupB}}
        {{item.name}}
      {{/sortable-item}}
    {{/each}}
  {{/sortable-group}}

{{/sortable-group}}
```

When you move an item from one list to the other, the two `onChange` handlers `updateListA` and `updateListB` would be invoked in the same run loop queue, receiving their respective new arrays.

Ideally the logic in an `onChange` handler should be no more complex than saying “this is what the list should look like now, cool or not cool?”.

So really it looks like the parent group’s role is to establish a relationship between the two sub-groups.

**Comment by rlivsey** [comment]
Agreed, maybe the parent doesn't need an `onChange` at all. 

In which case, does it make sense for that to maybe be a different component entirely?

**Comment by jgwhite** [comment]
@rlivsey perhaps, yeah.

```hbs
{{#sortable-manager as |manager|}}

  {{#sortable-group manager=manager onChange=(action updateListA) as |groupA|}}
    {{#each listA as |item|}}
      {{#sortable-item model=item group=groupA}}
        {{item.name}}
      {{/sortable-item}}
    {{/each}}
  {{/sortable-group}}

  {{#sortable-group manager=manager onChange=(action updateListB) as |groupB|}}
    {{#each listB as |item|}}
      {{#sortable-item model=item group=groupB}}
        {{item.name}}
      {{/sortable-item}}
    {{/each}}
  {{/sortable-group}}

{{/sortable-manager}}
```

**Comment by rlivsey** [comment]
Might have a chance to take a look at this tomorrow, just doing some more thinking out loud.

Should you have to drag an item over another list, or within some minimum distance, for it to be considered moved?

Eg consider two lists with 500px vertical space between them, if you drag an item from the bottom list ~300px upwards (so it's still ~200px below the top list) is that considered to now be part of the top list because it's now closer to the top one than the bottom?

**Comment by denzo** [comment]
I believe an item should be dragged over another list. I think https://jqueryui.com/sortable/#connect-lists has pretty solid functionality.

Also, just thinking out loud. In regards to wrapping sortable-group in sortable-manager, could that potentially be moved into a service ?

Consider the following scenario. Four lists on a page. The items can be moved between the lists 1 and 2. Items from list 1 and 2 can not be moved to lists 3 and 4. I think it would be beneficial to provide a group id to sortable-group and the service would determine if it is possible to move an item to a different list based on equality of group id.

**Comment by jgwhite** [comment]
@denzo I definitely like the idea of a separate logical service determining jumps between lists. Could you sketch out what all this might look like in action?

**Comment by denzo** [comment]
Apologies for the delayed response.

I have spent a few days understanding how the whole thing works. I must say @jgwhite has done an amazing job of putting together such a robust library.

I finally wrapped my head around it. I have forked this repo and created a working prototype. I have introduced sortable-manager service and moved some code around trying to focus on responsibilities of each class.

Apologies for tweaking your code and changing the styling but I needed that to get a full understanding.

Let me know what you think https://github.com/denzo/ember-sortable

**Comment by jgwhite** [comment]
Thanks @denzo. I’ll take a look as soon as I can.

**Comment by workmanw** [comment]
:+1: This would be really great.

**Comment by opsb** [comment]
@denzo it would be great to get this merged in, how far along is your fork? Does it need a little more work or is it ready to merge in? I'm pretty short on time this week but I might have some time to look at it over the weekend.

**Comment by denzo** [comment]
I have a working version https://github.com/denzo/ember-sortable so feel free to dig through. Run the demo (ember server) to see the behaviour.
* I'm away for the next couple of weeks and a limited access to the Internetz

**Comment by jgwhite** [comment]
Thanks again @denzo. I’m desperately short on time right now but super keen to look it over.

**Comment by TNAJanssen** [comment]
+1 what is the time line on this?

**Comment by opsb** [comment]
Depending on what you need there's still a little work left to do on @denzo's banch. I need scrollable lists which unfortunately makes things more complicated. If a div is scrollable then you can't drag outside unless the element you're dragging has position: fixed. I've had a play with that scenario, it's awkward because you need to use different co-ordinate systems, one relative to the list, another relative to the page (for the fixed position). The other option is to clone the dragged element and move that along with the original but with a position of fixed, in practice this seems to be working out better because the original mechanics are left untouched. Anyway, there's a little way to go here still I think, would be good to touch base with @denzo when he's back around.

**Comment by opsb** [comment]
I've now got a fork from @denzo's work at https://github.com/lytbulb/ember-sortable/commits/master which adds a couple of extra features that I need with multiple lists. Notably I've added support for scrollable lists, IE9+ and dropzones. If anyone wants to have a play it's functional and the couple of remaining glitches should be fixed this week. After that we're going to have somehow bring both forks into upstream...

**Comment by jgwhite** [comment]
Hey folks. Thanks all for looking into this. If anyone fancies making a PR, it’d make reviewing the changes easier.

**Comment by opsb** [comment]
There's a couple of wrinkles I want to iron before I start looking at a PR. Rebasing is also going to be fairly involved, @denzo forked off at 1.4.

**Comment by opsb** [comment]
It's also a spare-time only project for me so the timeline is a little difficult to call.

**Comment by jgwhite** [comment]
@opsb totally understand — this is spare-time only for me too :)

**Comment by TNAJanssen** [comment]
@opsb @jgwhite if you guys need help please tell me what i can do!

**Comment by jgwhite** [comment]
@TNAJanssen thanks. Essentially formualting the great work @denzo and @opsb have been doing into a PR. I really want to extract the goodness but really struggling to make time.

**Comment by opsb** [comment]
@TNAJanssen @jgwhite I've still got a list of about 5 outstanding bugs that I need to take care of before it's ready to go. I'm going to try and get those finished up over the weekend so the branch is fully functional. I'll also reduce the history down a bit so that a rebase is more manageable. I think that's the point when some help would be useful. If you could pick it up and rebase it against master that would be fantastic! If you do have some time to look at it I'd be very happy to answer any questions you've got, we can always set up a gitter room or jump on google hangouts. It sounds like @denzo will be around next week, I'm not sure how he's looking for time but maybe he'll want to run with it as well.

**Comment by rlivsey** [comment]
Anything I can do to help push this over the finish line? I can put aside some time next week.

**Comment by opsb** [comment]
@rlivsey sadly I'm not sure how salvageable this is. Master has moved forwards a fair bit and my work is forked from @denzo's fork which makes it difficult to integrate... I think someone is going to have to take a fresh crack it, I've already got my evenings/weekend time committed elsewhere so I won't have a chance to look at it. I'm very happy to get on a call with someone though and give them some background on what needs to be resolved.

**Comment by jgwhite** [comment]
@opsb thanks for your efforts. I’d be open to having a call next month some time. @rlivsey have been discussing possible next-generation APIs and I think the sorting-between-lists case should drive those conversations.

**Comment by lolmaus** [comment]
Any news one month later?

**Comment by jgwhite** [comment]
@lolmaus nothing to report so far. I'm personally struggling to find time to tackle any major features and also must admit that I don't actually *need* this particular one — though I'm sure I will on a future project.

Something that would be super helpful is for someone to add a nice complex example to the demo page so we all have a target in mind. Basically "I wish I could use ember-sortable exactly like this".

**Comment by lolmaus** [comment]
This example represents exactly what I need: https://jqueryui.com/sortable/#connect-lists

**Comment by jgwhite** [comment]
:+1:

**Comment by ianbale** [comment]
I have a closely related use case - currently implemented with jQuery sortable and jQuery draggable - but I'm having various problems with it which is why I went looking for an alternative and found your project...

I also need two lists. One contains my list of data items (main). This needs to be sortable. Then I have another list (pick list) from which I can drag into my main one. But, the difference is, I don't want items moved, I want them copied as I drag. The idea being that I can add any number of items of any type from my pick list into my main list in order to build that list up.

Here is a demo : https://jqueryui.com/resources/demos/draggable/sortable.html although it only has a single item in the "pick list", where as I have a sizeable list. In my case, the pick list does not need to be sortable. So maybe they can be created as a bunch of separate items which can be dragged into the list rather than as a list themselves...

So I guess the main functionality I need is the ability to drag a new item into the list.

I think my client on this project will be fine with me providing some assistance on this if it means we end up getting a working component which we can use for our project...

**Comment by jgwhite** [comment]
@ianbale that’s great information, many thanks. I must admit I’m at something of an impasse regarding this functionality. I’m not sure the existing API is quite right…

**Comment by opsb** [comment]
@jgwhite I think the current implementation would really benefit from a render buffer. It's been a while since I looked at the code so this is in broad strokes.

From
https://github.com/jgwhite/ember-sortable/blob/master/addon/components/sortable-group.js#L70

>   /**
    Prepare for sorting.
    Main purpose is to stash the current itemPosition so
    we don’t incur expensive re-layouts.
    @method prepare
  */

Effectively there's two render frames, the current positions, and the positions that will be rendered. Currently the two sets of information are kept alongside each other in different variables. If this information were extracted into a model then you could bind a copy of the model to the component and then make updates to a separate one which would be swapped in when it was ready to be rendered. 

I think the introduction of the service that @denzo used made a lot of sense as well. Potentially the introduction of a Drag model might make sense as an alternative as well. That would give you a place to keep track of the source/target etc.

Anyway, that's my 2 cents :)

**Comment by jgwhite** [comment]
Sounds y'all have some solid ideas for moving this forward. Who's up for a remote pair session in the new year?

**Comment by opsb** [comment]
@jgwhite Let's see if we can something up. I'll also likely be over in zap towers at some point in jan so we could maybe do a session there.

**Comment by jgwhite** [comment]
@opsb that'd be very awesome

**Comment by ianbale** [comment]
Just a quick comment on the parent/child sortable idea. It would be better, if possible, to avoid having to nest the two lists you want to drag between inside a single tag. The reason being that there will be many use cases where one list will be contained within one component and the other within a second, maybe nested quite deeply within a complex application.

I am thinking of my case again... but this is one one of several projects with the same sort of structure, so I doubt very much that my cases are unique.

This sort of thing can easily be achieved with jQuery draggable, where the two lists can be placed anywhere. It would be nice if this was just as flexible.

I've integrated what you have so far into my project and most of the problems I had with jQuery sortable have gone away. Very impressive! :-) So that means that I really need to get something working as quickly as possible to be able to drag a new item into the list (It does not need to be another list. I just need to be able to drag in a new [pseudo] item via a drag mechanism). Do you think this should be a part of this enhancement to drag between lists, or shall I open a new issue for this? I'm definitely up for working on this with you, but have a few questions about how it works. I think I've worked out most of it, but some bits are still eluding me...

**Comment by myared** [comment]
Just checking in on this issue as I may be in need of this feature soon. I wanted to see if anyone is still currently working on it. If not, I may take a stab.

**Comment by jgwhite** [comment]
@myared I’ve got some sponsorship to work on this feature next week thanks to the good folks at [ScreenSteps](http://www.screensteps.com/). Watch this space :)

**Comment by jheth** [comment]
I have a need to sort between lists as well as sort the groupings of lists. Would the work being done here support that case?

For example, I want to sort Section 1 <=> Section 2 and then any page in or between sections.

```html
<ul class="sections">
   <li class="section">Section 1
        <ul class="pages">
             <li>Page 1</li>
             <li>Page 2</li>
        </ul>
   </li>
   <li class="section">Section 2
        <ul class="pages">
             <li>Page 3</li>
             <li>Page 3</li>
        </ul>
   </li>
</ul>
```

**Comment by jgwhite** [comment]
@jheth yes, for sure

**Comment by opsb** [comment]
@jgwhite if it's useful I'd be happy to talk you through my experience of implementing it. There's a few gotchas.

**Comment by jgwhite** [comment]
@opsb that'd be very useful. Works commences next Tuesday. I'll give you a shout once I've gotten the lay of the land.

**Comment by opsb** [comment]
:+1:

**Comment by onlymejosh** [comment]
Any news on this @jgwhite :)

**Comment by jgwhite** [comment]
@onlymejosh check the jj-abrams branch for progress.

**Comment by onlymejosh** [comment]
Looks like its going pretty well, I cant really tell whats left to do? Happy to help

**Comment by scottmessinger** [comment]
@jgwhite I pulled your jj-abrams branch and got it working in my app! Everytime I've hacked on ember-sortable, I find it really pleasant. I find reading your code and following the logic quite easy -- thanks so much for doing this. 

As I working with it, there were just a couple things I ran into that I changed:
- I changed the translates to translate3d and debounced them. Pretty simple change
- the `findParent` code will stop if the parent can't receive a sortable **even if it's child can**. I wrote the code to flatten out the tree. Is that a change you'd be interested in seeing a PR for?
- I needed more params sent in the `onMove` callback. I wasn't sure how to fix this. The code sends back the `newParent` which is helpful, but I need to know the newParent's model. For instance, I'm moving sections in a lesson plan, I need to know the new sections array as well as the parent lesson document. I ended up doing this:

```js
let newParentContainer = receiver.sortableModelContainer;
this.sendAction('onMove', model, { oldParent, newParent, newParentContainer, position });
```

I don't like the name and it feels like the more fundamental problem is letting the app developer choose arbitrary number of things to send in the onMove callback. Any ideas about that?

Thanks again so much,
Scott

**Comment by jgwhite** [comment]
> I changed the translates to translate3d and debounced them. Pretty simple change

Sounds like it's worth a PR :) Does it have a noticeable effect on performance?

> the findParent code will stop if the parent can't receive a sortable even if it's child can. I wrote the code to flatten out the tree. Is that a change you'd be interested in seeing a PR for?

Yes definitely. That logic is in its very early stages in that branch. I tried a number of approaches to walking the tree (including quad trees, believe it or not) but kept coming back to this approach. The check whether a child can receive a model was a last minute addition.

> I needed more params sent in onMove…

This is interesting. It has always bothered me that the APIs for ember-sortable aren't very well aligned with native dragdrop APIs, where you can indeed pass arbitrary payloads around. I don't see a downside in leaving it open to extension but I wouldn't want consumer of the library to need to be aware of too much extra context when handling an onMove.

Could you assign a more complex model to that node from the outset?

Also, you say you needed newParent's model. Remind me why you can't call `newParent.get('sortableModel')` (I may be forgetting something obvious).

**Comment by shanebruggeman** [comment]
Wow, lots of great discussion on this threat. Is this still in progress? I'd love to be able to transfer items between lists

**Comment by jgwhite** [comment]
As of today, this isn’t under active development. This thread has been terrific but I’m going to close it and hopefully some brave soul can use it as a source of inspiration. Perhaps we’ll see ember-power-sortable in the future?
